src/libexpr/eval.cc

```diff
@@ -806,14 +806,16 @@ void EvalState::runDebugRepl(const Error * error, const Env & env, const Expr &
     }
 }
 
-void EvalState::addErrorTrace(Error & e, const char * s, const std::string & s2) const
+template<typename... Args>
+void EvalState::addErrorTrace(Error & e, const Args & ... formatArgs) const
 {
-    e.addTrace(nullptr, s, s2);
+    e.addTrace(nullptr, HintFmt(formatArgs...));
 }
 
-void EvalState::addErrorTrace(Error & e, const PosIdx pos, const char * s, const std::string & s2, bool frame) const
+template<typename... Args>
+void EvalState::addErrorTrace(Error & e, const PosIdx pos, const Args & ... formatArgs) const
 {
-    e.addTrace(positions[pos], HintFmt(s, s2), frame);
+    e.addTrace(positions[pos], HintFmt(formatArgs...));
 }
```

Comment:
This makes the addition of optional arguments more invasive.
I'm proposing to add one in
- https://github.com/NixOS/nix/pull/8623

Reply:
I think it's very clumsy to have a mandatory "formatting parameter" argument, and we pass in empty strings here. To add more optional arguments, let's add an overload that takes a `HintFmt` explicitly. (I can add one here as well.)

Reply:
I agree `s2` is very clumsy. We could revisit the varargs later if needed. Not even many call sites are affected anyway.